### PR TITLE
[ fix #1128 ] Use the BUILTIN Maybe

### LINKED ISSUE #1128: Use the BUILTIN Maybe
Following https://github.com/agda/agda/commit/1eec63b1c5566b252c0a4a815ce1df99a772c475 we have a BUILTIN maybe type constructor.
We should re-export this one rather than defining our own. We can then get rid
of the support for `Maybe` in `Foreign/Haskell`.

This work should be done on the **experimental** branch.

### PR DESCRIPTION
**Pull request by gallais**

### CODE REVIEW

**Comment by MatthewDaggitt** [comment]
Looks good! CHANGELOG entry?

**Comment by gallais** [comment]
I am  getting an error:

```agda
Failed to solve the following constraints:
  Resolve instance argument
    _97
      : {A.a : Level} {A = A₁ : Set A.a} →
        Coercible (List (Maybe _A_99) → List _A_99)
        (List (Maybe A₁) → List A₁)
  Candidates
    coerce-fun :
      {A.a : Level} {A = A₁ : Set A.a} {B.b : Level} {B : Set B.b}
      {C.c : Level} {C : Set C.c} {D.d : Level} {D : Set D.d}
      ⦃ _ : Coercible A₁ B ⦄ ⦃ _ : Coercible C D ⦄ →
      Coercible (B → C) (A₁ → D)
    coerce-refl : {A.a : Level} {A = A₁ : Set A.a} → Coercible A₁ A₁
```

But taking any of the two would yield the same solution: `TrustMe` so it
should not matter whether we are getting the first or the second solution.
Opinion @UlfNorell?

**Comment by UlfNorell** [comment]
Surely that would depend on the choices of coerce instances for the domain and codomain? Instances are checked for equality before solving recursive goals, if I'm not mistaken.

cc @jespercockx

**Comment by gallais** [comment]
Everything is eventually just `TrustMe`: http://agda.github.io/agda-stdlib/Foreign.Haskell.Coerce.html

**Comment by UlfNorell** [comment]
Hm, that's strange then. Something else must be preventing the search from proceeding.

**Comment by jespercockx** [comment]
When exactly are you getting this error? I'm trying to reproduce it so I can take a look at the debug output.

**Comment by gallais** [comment]
@jespercockx It's in README.Foreign.Haskell on the branch for this PR.
Cf. https://travis-ci.org/github/agda/agda-stdlib/jobs/692191061

**Comment by jespercockx** [comment]
From the debug output, it seems that the first candidate is a "fresh meta" (i.e. one that was created during checking of the candidate) rather than the actual term `TrustMe`. This might just be a case of a missing call to `instantiate`.

**Comment by jespercockx** [comment]
There actually *is* a call to `instantiateFull` on the term before the call to `dropSameCandidates`, so now I have no idea what's actually going on. But I think it's a safe bet that this is a bug in Agda itself and not with how you're using the instance search.

**Comment by jespercockx** [comment]
The issue in Agda has been fixed now.

**Comment by gallais** [comment]
There's a new one (for me): it seems that I have forgotten to automatically
attach compilation pragmas to the builtin `Maybe` like we have for `List`.
So I'll need to fix that first.

**Comment by gallais** [comment]
I have fixed the compilation problem but now there are issues with the
termination checker. :(

**Comment by jespercockx** [comment]
You could try to fix this by adding some `{-# INLINE #-}` pragmas (see https://github.com/agda/agda/issues/4681).

**Comment by gallais** [comment]
You cannot inline anonymous pattern-matching functions! I have filed agda/agda#4702

File: src/Foreign/Haskell/Maybe.agda [hunk 1]
```diff
@@ -1,38 +0,0 @@
...
-
-private
-  variable
-    a : Level
-    A : Set a
-
-------------------------------------------------------------------------
--- Definition
-
-data Maybe (A : Set a) : Set a where
```
Comment (MatthewDaggitt) [inline src/Foreign/Haskell/Maybe.agda:22]: For backwards compatibility we should really keep this module, but mark the internal terms as deprecated and have the warning explain they're no longer needed?

**Pull request merged**